source/common/network/connection_impl.cc

```diff
@@ -115,15 +115,15 @@ void ConnectionImpl::close(ConnectionCloseType type) {
 
   uint64_t data_to_write = write_buffer_->length();
   ENVOY_CONN_LOG(debug, "closing data_to_write={} type={}", *this, data_to_write, enumToInt(type));
+  if (data_to_write > 0) {
+    // Try to write as much as we can if there is pending data.
+    transport_socket_->doWrite(*write_buffer_, type == ConnectionCloseType::NoFlush ||
+                                                   !transport_socket_->canFlushClose());
+  }
+
```

Comment:
Let's start with the usual feedback: please add some tests.

Going beyond that, I'm not 100% sure it makes sense to attempt to flush before close in all cases.  It's possible that we don't really mean it when we currently say ConnectionCloseType::NoFlush given that the previous version of this method attempted a write when type was NoFlush.  Adding a NoReallyThrowDataAwayAndClose enum type seems like overkill, but it also seems somewhat appropriate.

Reply:
Oh, I didn't indent to merge this PR as-is, I've added it here to start the discussion about possible solution for #13890. A test for it would be working `FlushWrite` in #13858, but that's not merged yet.

At this point I'm not really sure what the current semantics are and why `NoFlush` does flush data, to be honest.